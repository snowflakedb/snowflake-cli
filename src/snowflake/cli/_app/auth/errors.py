class OidcProviderError(Exception):
    """Base exception for OIDC provider related errors."""

    ...


class OidcProviderNotFoundError(OidcProviderError):
    """Exception raised when requested OIDC provider is not found or unknown."""

    ...


class OidcProviderUnavailableError(OidcProviderError):
    """Exception raised when OIDC provider is not available in current environment."""

    ...


class OidcProviderAutoDetectionError(OidcProviderError):
    """Exception raised when auto-detection of OIDC provider fails."""

    ...


class OidcTokenRetrievalError(OidcProviderError):
    """Exception raised when OIDC token cannot be retrieved."""

    ...
