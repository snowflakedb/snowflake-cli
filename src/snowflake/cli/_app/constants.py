# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

from typing import Literal

PARAM_APPLICATION_NAME: Literal["snowcli"] = "snowcli"

# This is also defined on server side. Changing this parameter would require
# a change in https://github.com/snowflakedb/snowflake
INTERNAL_APPLICATION_NAME: Literal["SNOWFLAKE_CLI"] = "SNOWFLAKE_CLI"

# Authenticator types
AUTHENTICATOR_WORKLOAD_IDENTITY: Literal["WORKLOAD_IDENTITY"] = "WORKLOAD_IDENTITY"
AUTHENTICATOR_SNOWFLAKE_JWT: Literal["SNOWFLAKE_JWT"] = "SNOWFLAKE_JWT"
AUTHENTICATOR_USERNAME_PASSWORD_MFA: Literal[
    "username_password_mfa"
] = "username_password_mfa"
AUTHENTICATOR_OAUTH_AUTHORIZATION_CODE: Literal[
    "OAUTH_AUTHORIZATION_CODE"
] = "OAUTH_AUTHORIZATION_CODE"
