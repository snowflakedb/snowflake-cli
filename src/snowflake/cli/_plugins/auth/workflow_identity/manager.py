# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import Optional

from click import ClickException
from snowflake.cli.api.cli_global_context import get_cli_context
from snowflake.cli.api.console import cli_console
from snowflake.cli.api.sql_execution import SqlExecutionMixin
from snowflake.connector.cursor import SnowflakeCursor


class WorkflowIdentityManager(SqlExecutionMixin):
    """
    Manager for GitHub workflow identity federation authentication.
    """

    def setup(self, github_repository: str) -> None:
        """
        Sets up GitHub workflow identity federation for the specified repository.

        Args:
            github_repository: GitHub repository in format 'owner/repo'
        """
        if not self._validate_github_repository_format(github_repository):
            raise ClickException(
                "Invalid GitHub repository format. Expected format: 'owner/repo'"
            )

        cli_context = get_cli_context()
        user = cli_context.connection.user

        # Create or update the security integration for GitHub workflow identity federation
        self._create_security_integration(github_repository)
        cli_console.step(f"Created security integration for {github_repository}")

        # Set the workflow identity property for the user
        self._set_workflow_identity_property(user, github_repository)
        cli_console.step(f"Set workflow identity property for user {user}")

    def status(self) -> str:
        """
        Returns the status of GitHub workflow identity federation configuration.

        Returns:
            Status message indicating if workflow identity federation is configured
        """
        cli_context = get_cli_context()
        user = cli_context.connection.user

        try:
            # Check if workflow identity is configured for the user
            workflow_identity = self._get_workflow_identity_property(user)
            if workflow_identity:
                return f"Configured for repository: {workflow_identity}"
            else:
                return "Not configured"
        except Exception as e:
            return f"Error checking status: {str(e)}"

    def remove(self) -> SnowflakeCursor:
        """
        Removes the GitHub workflow identity federation configuration.

        Returns:
            SnowflakeCursor with the result of the operation
        """
        cli_context = get_cli_context()
        user = cli_context.connection.user

        return self.execute_query(f"ALTER USER {user} UNSET GITHUB_WORKFLOW_IDENTITY")

    def _validate_github_repository_format(self, repository: str) -> bool:
        """
        Validates that the GitHub repository is in the correct format (owner/repo).

        Args:
            repository: The repository string to validate

        Returns:
            True if format is valid, False otherwise
        """
        parts = repository.split("/")
        return len(parts) == 2 and all(part.strip() for part in parts)

    def _create_security_integration(self, github_repository: str) -> SnowflakeCursor:
        """
        Creates or replaces a security integration for GitHub workflow identity federation.

        Args:
            github_repository: GitHub repository in format 'owner/repo'

        Returns:
            SnowflakeCursor with the result of the operation
        """
        integration_name = f"GITHUB_WIF_{github_repository.replace('/', '_').upper()}"

        sql = f"""
        CREATE OR REPLACE SECURITY INTEGRATION {integration_name}
        TYPE = OAUTH
        OAUTH_CLIENT = GITHUB_ACTIONS
        OAUTH_ALLOWED_AUTHORIZATION_ENDPOINTS = ('https://token.actions.githubusercontent.com')
        OAUTH_ALLOWED_TOKEN_ENDPOINTS = ('https://token.actions.githubusercontent.com')
        OAUTH_ALLOWED_SCOPES = ('repository:read')
        OAUTH_ISSUER = 'https://token.actions.githubusercontent.com'
        OAUTH_AUDIENCE = '{github_repository}'
        """

        return self.execute_query(sql)

    def _set_workflow_identity_property(
        self, user: str, github_repository: str
    ) -> SnowflakeCursor:
        """
        Sets the GitHub workflow identity property for a user.

        Args:
            user: The Snowflake user name
            github_repository: GitHub repository in format 'owner/repo'

        Returns:
            SnowflakeCursor with the result of the operation
        """
        return self.execute_query(
            f"ALTER USER {user} SET GITHUB_WORKFLOW_IDENTITY = '{github_repository}'"
        )

    def _get_workflow_identity_property(self, user: str) -> Optional[str]:
        """
        Gets the GitHub workflow identity property for a user.

        Args:
            user: The Snowflake user name

        Returns:
            The workflow identity property value if set, None otherwise
        """
        cursor = self.execute_query(f"DESCRIBE USER {user}")
        properties = cursor.fetchall()

        for prop in properties:
            if prop[0] == "GITHUB_WORKFLOW_IDENTITY" and prop[1]:
                return prop[1]

        return None
