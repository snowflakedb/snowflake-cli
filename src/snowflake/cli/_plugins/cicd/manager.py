import enum
from typing import List, Optional, Type

from snowflake.cli.api.secure_path import SecurePath


class CIProviderChoices(str, enum.Enum):
    GITHUB = "GITHUB"
    GITLAB = "GITLAB"


class CIProvider:
    NAME: str

    @classmethod
    def cleanup(cls, root: SecurePath) -> None:
        raise NotImplementedError()

    @classmethod
    def from_choice(cls, choice: CIProviderChoices) -> "CIProvider":
        return {
            GithubProvider.NAME: GithubProvider,
            GitLabProvider.NAME: GitLabProvider,
        }[choice.name]()

    @classmethod
    def all(cls) -> List[Type["CIProvider"]]:  # noqa: A003
        return [GithubProvider, GitLabProvider]

    def has_template(self, root_dir: SecurePath) -> bool:
        raise NotImplementedError()

    def copy(self, source: SecurePath, destination: SecurePath):
        raise NotImplementedError()


class GithubProvider(CIProvider):
    NAME = CIProviderChoices.GITHUB.name

    @classmethod
    def cleanup(cls, root_dir: SecurePath):
        (root_dir / ".github").rmdir(recursive=True)

    def has_template(self, root_dir: SecurePath) -> bool:
        return (root_dir / ".github/workflows").exists()

    def copy(self, source: SecurePath, destination: SecurePath) -> None:
        (source / ".github").copy(destination.path, dirs_exist_ok=True)


class GitLabProvider(CIProvider):
    NAME = CIProviderChoices.GITLAB.name

    @classmethod
    def cleanup(cls, root_dir: SecurePath):
        (root_dir / ".gitlab-ci.yml").unlink(missing_ok=True)

    def has_template(self, root_dir: SecurePath) -> bool:
        return (root_dir / ".gitlab-ci.yml").exists()

    def copy(self, source: SecurePath, destination: SecurePath) -> None:
        if (destination / ".gitlab-ci.yml").exists():
            (destination / ".gitlab-ci.yml").unlink()
        (source / ".gitlab-ci.yml").move(destination.path)


class CIProviderManager:
    @staticmethod
    def project_post_gen_cleanup(
        selected_provider: Optional[CIProvider], template_root: SecurePath
    ):
        for provider_cls in CIProvider.all():
            if selected_provider and not isinstance(selected_provider, provider_cls):
                provider_cls.cleanup(template_root)
