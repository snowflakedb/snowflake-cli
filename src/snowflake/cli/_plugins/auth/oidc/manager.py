# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import logging
from typing import TypeAlias

from snowflake.cli._app.auth.errors import OidcProviderError
from snowflake.cli._app.auth.oidc_providers import (
    OidcProviderType,
    OidcProviderTypeWithAuto,
    auto_detect_oidc_provider,
    get_active_oidc_provider,
)
from snowflake.cli.api.exceptions import CliError

logger = logging.getLogger(__name__)


Providers: TypeAlias = OidcProviderType | OidcProviderTypeWithAuto


class OidcManager:
    """
    Manages OIDC authentication.

    This class provides methods to read OIDC configurations for authentication.
    """

    def read_token(
        self,
        provider_type: Providers = OidcProviderTypeWithAuto.AUTO,
    ) -> str:
        """
        Reads OIDC token based on the specified provider type.

        Args:
            provider_type: Type of provider to read token from ("auto" for auto-detection)

        Returns:
            Token string or provider information

        Raises:
            CliError: If token reading fails
        """
        logger.info("Reading OIDC token with provider type: %s", provider_type)

        try:
            if provider_type == OidcProviderTypeWithAuto.AUTO:
                provider = auto_detect_oidc_provider()
            else:
                provider = get_active_oidc_provider(provider_type.value)
            return provider.get_token()
        except OidcProviderError as e:
            logger.error("OIDC provider error: %s", str(e))
            raise CliError(str(e))
