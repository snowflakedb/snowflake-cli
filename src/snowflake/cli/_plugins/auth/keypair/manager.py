from enum import Enum
from typing import Dict, List, Optional, Tuple

from click import ClickException
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric import rsa
from snowflake.cli._plugins.object.manager import ObjectManager
from snowflake.cli.api import exceptions
from snowflake.cli.api.cli_global_context import (
    _CliGlobalContextAccess,
    get_cli_context,
)
from snowflake.cli.api.config import (
    connection_exists,
    get_connection_dict,
    set_config_value,
)
from snowflake.cli.api.console import cli_console
from snowflake.cli.api.constants import ObjectType
from snowflake.cli.api.identifiers import FQN
from snowflake.cli.api.secret import SecretType
from snowflake.cli.api.secure_path import SecurePath
from snowflake.cli.api.sql_execution import SqlExecutionMixin
from snowflake.connector import DictCursor
from snowflake.connector.cursor import SnowflakeCursor


class PublicKeyProperty(Enum):
    RSA_PUBLIC_KEY = "RSA_PUBLIC_KEY"
    RSA_PUBLIC_KEY_2 = "RSA_PUBLIC_KEY_2"


class AuthManager(SqlExecutionMixin):
    def setup(
        self,
        connection_name: Optional[str],
        key_length: int,
        output_path: SecurePath,
        private_key_passphrase: SecretType,
    ):
        # When the user provide new connection name
        if connection_name and connection_exists(connection_name):
            raise ClickException(
                f"Connection with name {connection_name} already exists."
            )

        cli_context = get_cli_context()
        # When the use not provide connection name, so we overwrite the current connection
        if not connection_name:
            connection_name = cli_context.connection_context.connection_name

        key_name = AuthManager._get_free_key_name(output_path, connection_name)  # type: ignore[arg-type]
        self._generate_key_pair_and_set_public_key(
            user=cli_context.connection.user,
            key_length=key_length,
            output_path=output_path,
            key_name=key_name,  # type: ignore[arg-type]
            private_key_passphrase=private_key_passphrase,
        )

        self._create_or_update_connection(
            current_connection=cli_context.connection_context.connection_name,
            connection_name=connection_name,  # type: ignore[arg-type]
            private_key_path=self._get_private_key_path(
                output_path=output_path, key_name=key_name  # type: ignore[arg-type]
            ),
        )

    def rotate(
        self,
        key_length: int,
        output_path: SecurePath,
        private_key_passphrase: SecretType,
    ):
        cli_context = get_cli_context()
        connection_name = cli_context.connection_context.connection_name

        self._ensure_connection_has_private_key(
            cli_context.connection_context.connection_name
        )

        public_key, public_key_2 = self._get_public_keys()

        if not public_key and not public_key_2:
            raise ClickException("No public key found. Use the setup command first.")

        if public_key_2:
            self.set_public_key(
                cli_context.connection.user,
                PublicKeyProperty.RSA_PUBLIC_KEY,
                public_key_2,
            )

        key_name = AuthManager._get_free_key_name(output_path, connection_name)
        public_key = self._generate_keys_and_return_public_key(
            key_length=key_length,
            output_path=output_path,
            key_name=key_name,
            private_key_passphrase=private_key_passphrase,
        )
        self.set_public_key(
            cli_context.connection.user, PublicKeyProperty.RSA_PUBLIC_KEY_2, public_key
        )
        self._create_or_update_connection(
            current_connection=cli_context.connection_context.connection_name,
            connection_name=connection_name,
            private_key_path=self._get_private_key_path(
                output_path=output_path, key_name=key_name
            ),
        )

    def _generate_key_pair_and_set_public_key(
        self,
        user: str,
        key_length: int,
        output_path: SecurePath,
        key_name: str,
        private_key_passphrase: SecretType,
    ):
        public_key_exists, public_key_2_exists = self._get_public_keys()

        if public_key_exists or public_key_2_exists:
            raise exceptions.CouldNotSetKeyPairError()

        if not output_path.exists():
            output_path.mkdir(parents=True)

        public_key = self._generate_keys_and_return_public_key(
            key_length=key_length,
            output_path=output_path,
            key_name=key_name,  # type: ignore[arg-type]
            private_key_passphrase=private_key_passphrase,
        )
        self.set_public_key(user, PublicKeyProperty.RSA_PUBLIC_KEY, public_key)

    def list_keys(self) -> List[Dict]:
        key_properties = [
            "RSA_PUBLIC_KEY",
            "RSA_PUBLIC_KEY_FP",
            "RSA_PUBLIC_KEY_LAST_SET_TIME",
            "RSA_PUBLIC_KEY_2",
            "RSA_PUBLIC_KEY_2_FP",
            "RSA_PUBLIC_KEY_2_LAST_SET_TIME",
        ]
        cursor = ObjectManager(connection=self._conn).describe(
            object_type=ObjectType.USER.value.sf_name,
            fqn=FQN.from_string(self._conn.user),
            cursor_class=DictCursor,
        )
        only_public_key_properties = [
            p for p in cursor.fetchall() if p.get("property") in key_properties
        ]
        return only_public_key_properties

    def set_public_key(
        self, user: str, public_key_property: PublicKeyProperty, public_key: str
    ) -> SnowflakeCursor:
        return self.execute_query(
            f"ALTER USER {user} SET {public_key_property.value}='{public_key}'"
        )

    def remove_public_key(
        self, public_key_property: PublicKeyProperty
    ) -> SnowflakeCursor:
        cli_context = get_cli_context()
        return self.execute_query(
            f"ALTER USER {cli_context.connection.user} UNSET {public_key_property.value}"
        )

    def status(self):
        cli_context = get_cli_context()
        self._ensure_connection_has_private_key(
            cli_context.connection_context.connection_name
        )
        cli_console.step("Private key set for connection - OK")
        self._check_connection(cli_context)
        cli_console.step("Test connection - OK")

    def extend_connection_add(
        self,
        connection_name: str,
        connection_options: Dict,
        key_length: int,
        output_path: SecurePath,
        private_key_passphrase: SecretType,
    ) -> Dict:
        key_name = AuthManager._get_free_key_name(output_path, connection_name)

        self._generate_key_pair_and_set_public_key(
            user=connection_options["user"],
            key_length=key_length,
            output_path=output_path,
            key_name=key_name,
            private_key_passphrase=private_key_passphrase,
        )

        connection_options["authenticator"] = "SNOWFLAKE_JWT"
        connection_options["private_key_file"] = str(
            self._get_private_key_path(output_path=output_path, key_name=key_name).path
        )
        if connection_options.get("password"):
            del connection_options["password"]

        return connection_options

    @staticmethod
    def _ensure_connection_has_private_key(connection_name: str) -> None:
        connection = get_connection_dict(connection_name)
        if not connection.get("private_key_file") and not connection.get(
            "private_key_path"
        ):
            raise ClickException(
                f"The private key is not set in {connection_name} connection."
            )

    @staticmethod
    def _check_connection(cli_context: _CliGlobalContextAccess) -> None:
        cli_context.connection

    def _get_public_keys(self) -> Tuple[str, str]:
        keys = self.list_keys()
        public_key = ""
        public_key_2 = ""
        for p in keys:
            if (
                p.get("property") == PublicKeyProperty.RSA_PUBLIC_KEY.value
                and p.get("value") != "null"
            ):
                public_key = p.get("value")  # type: ignore
            if (
                p.get("property") == PublicKeyProperty.RSA_PUBLIC_KEY_2.value
                and p.get("value") != "null"
            ):
                public_key_2 = p.get("value")  # type: ignore
        return public_key, public_key_2

    @staticmethod
    def _generate_keys_and_return_public_key(
        key_length: int,
        output_path: SecurePath,
        key_name: str,
        private_key_passphrase: SecretType,
    ) -> str:
        private_key = rsa.generate_private_key(
            public_exponent=65537,
            key_size=key_length,
        )

        if private_key_passphrase:
            pem = SecretType(
                private_key.private_bytes(
                    encoding=serialization.Encoding.PEM,
                    format=serialization.PrivateFormat.PKCS8,
                    encryption_algorithm=serialization.BestAvailableEncryption(
                        private_key_passphrase.value.encode("utf-8")
                    ),
                )
            )
            cli_console.message(
                "Set the `PRIVATE_KEY_PASSPHRASE` environment variable before using the connection."
            )
        else:
            pem = SecretType(
                private_key.private_bytes(
                    encoding=serialization.Encoding.PEM,
                    format=serialization.PrivateFormat.PKCS8,
                    encryption_algorithm=serialization.NoEncryption(),
                )
            )

        with AuthManager._get_private_key_path(output_path, key_name).open(
            mode="wb"
        ) as file:
            file.write(pem.value)

        public_key = private_key.public_key()
        public_pem = public_key.public_bytes(
            encoding=serialization.Encoding.PEM,
            format=serialization.PublicFormat.SubjectPublicKeyInfo,
        )
        with AuthManager._get_public_key_path(output_path, key_name).open(
            mode="wb"
        ) as file:
            file.write(public_pem)

        return public_pem.decode("utf-8")

    @staticmethod
    def _get_free_key_name(output_path: SecurePath, key_name: str) -> str:
        new_private_key = f"{key_name}.p8"
        new_public_key = f"{key_name}.pub"
        new_key_name = key_name
        counter = 1

        while (
            (output_path / new_private_key).exists()
            and (output_path / new_public_key).exists()
            and counter <= 100
        ):
            new_key_name = f"{key_name}_{counter}"
            new_private_key = f"{new_key_name}.p8"
            new_public_key = f"{new_key_name}.pub"
            counter += 1

        if counter == 100:
            raise ClickException(
                "Too many key pairs with the same name in the output directory."
            )

        return new_key_name

    @staticmethod
    def _get_private_key_path(output_path: SecurePath, key_name: str) -> SecurePath:
        return (output_path / f"{key_name}.p8").resolve()

    @staticmethod
    def _get_public_key_path(output_path: SecurePath, key_name: str) -> SecurePath:
        return (output_path / f"{key_name}.pub").resolve()

    @staticmethod
    def _create_or_update_connection(
        current_connection: Optional[str],
        connection_name: str,
        private_key_path: SecurePath,
    ):
        connection = get_connection_dict(current_connection)
        connection.pop("password", None)
        connection["authenticator"] = "SNOWFLAKE_JWT"
        connection["private_key_file"] = str(private_key_path.path)

        set_config_value(["connections", connection_name], value=connection)
