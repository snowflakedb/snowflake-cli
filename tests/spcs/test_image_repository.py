# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import json
from typing import Dict
from unittest import mock
from unittest.mock import Mock, patch

import pytest
from snowflake.cli._plugins.spcs.image_repository.manager import ImageRepositoryManager
from snowflake.cli.api.constants import ObjectType
from snowflake.cli.api.exceptions import (
    DatabaseNotProvidedError,
    SchemaNotProvidedError,
)
from snowflake.connector.cursor import SnowflakeCursor
from snowflake.connector.errors import ProgrammingError

from tests.spcs.test_common import SPCS_OBJECT_EXISTS_ERROR

MOCK_ROWS = [
    [
        "2023-01-01 00:00:00",
        "IMAGES",
        "DB",
        "SCHEMA",
        "orgname-alias.registry.snowflakecomputing.com/DB/SCHEMA/IMAGES",
        "ROLE",
        "ROLE",
        "",
    ]
]

MOCK_COLUMNS = [
    "created_on",
    "name",
    "database_name",
    "schema_name",
    "repository_url",
    "owner",
    "owner_role_type",
    "comment",
]
MOCK_ROWS_DICT = [
    {col_name: col_val for col_name, col_val in zip(MOCK_COLUMNS, row)}
    for row in MOCK_ROWS
]


@pytest.mark.parametrize(
    "replace, if_not_exists, expected_query",
    [
        (False, False, "create image repository test_repo"),
        (False, True, "create image repository if not exists test_repo"),
        (True, False, "create or replace image repository test_repo"),
        # (True, True) is an invalid case as OR REPLACE and IF NOT EXISTS are mutually exclusive.
    ],
)
@mock.patch(
    "snowflake.cli._plugins.spcs.image_repository.manager.ImageRepositoryManager._execute_schema_query"
)
def test_create(mock_execute, replace, if_not_exists, expected_query):
    repo_name = "test_repo"
    cursor = Mock(spec=SnowflakeCursor)
    mock_execute.return_value = cursor
    result = ImageRepositoryManager().create(
        name=repo_name, replace=replace, if_not_exists=if_not_exists
    )
    mock_execute.assert_called_once_with(expected_query, name=repo_name)
    assert result == cursor


def test_create_replace_and_if_not_exist():
    with pytest.raises(ValueError) as e:
        ImageRepositoryManager().create(
            name="test_repo", replace=True, if_not_exists=True
        )
    assert "mutually exclusive" in str(e.value)


@mock.patch(
    "snowflake.cli._plugins.spcs.image_repository.manager.ImageRepositoryManager.create"
)
def test_create_cli(mock_create, mock_cursor, runner, os_agnostic_snapshot):
    repo_name = "test_repo"
    cursor = mock_cursor(
        rows=[[f"Image Repository {repo_name.upper()} successfully created."]],
        columns=["status"],
    )
    mock_create.return_value = cursor
    command = ["spcs", "image-repository", "create", repo_name]
    result = runner.invoke(command)
    mock_create.assert_called_once_with(
        name=repo_name, replace=False, if_not_exists=False
    )
    assert result.exit_code == 0, result.output
    assert result.output == os_agnostic_snapshot


def test_create_cli_replace_and_if_not_exists_fails(runner, os_agnostic_snapshot):
    command = [
        "spcs",
        "image-repository",
        "create",
        "test_repo",
        "--replace",
        "--if-not-exists",
    ]
    result = runner.invoke(command)
    assert result.exit_code == 2
    assert result.output == os_agnostic_snapshot


@mock.patch(
    "snowflake.cli._plugins.spcs.image_repository.manager.ImageRepositoryManager._execute_schema_query"
)
@mock.patch(
    "snowflake.cli._plugins.spcs.image_repository.manager.handle_object_already_exists"
)
def test_create_repository_already_exists(mock_handle, mock_execute):
    repo_name = "test_object"
    mock_execute.side_effect = SPCS_OBJECT_EXISTS_ERROR
    ImageRepositoryManager().create(repo_name, replace=False, if_not_exists=False)
    mock_handle.assert_called_once_with(
        SPCS_OBJECT_EXISTS_ERROR,
        ObjectType.IMAGE_REPOSITORY,
        repo_name,
        replace_available=True,
    )


@patch(
    "snowflake.cli._plugins.spcs.image_repository.commands.ImageRepositoryManager.list_images"
)
def test_list_images_cli(
    mock_list_images,
    runner,
    mock_cursor,
):
    cursor = mock_cursor(
        rows=[
            [
                "2024-10-11 14:23:49-07:00",
                "echo_service",
                "latest",
                "sha256:a8a001fef406fdb3125ce8e8bf9970c35af7084",
                "/db/schema/repo/echo_service",
            ]
        ],
        columns=["created_on", "image_name", "tags", "digest", "image_path"],
    )
    mock_list_images.return_value = cursor

    result = runner.invoke(
        ["spcs", "image-repository", "list-images", "IMAGES", "--format", "JSON"]
    )

    assert result.exit_code == 0, result.output
    assert "/db/schema/repo/echo_service" in result.output


@patch(
    "snowflake.cli._plugins.spcs.image_repository.commands.ImageRepositoryManager.execute_query"
)
def test_list_images(mock_execute_query):
    repo_name = "test_repo"
    cursor = Mock(spec=SnowflakeCursor)
    mock_execute_query.return_value = cursor
    result = ImageRepositoryManager().list_images(repo_name)
    expected_query = f"show images in image repository test_repo"
    mock_execute_query.assert_called_once_with(expected_query)
    assert result == cursor


@mock.patch("snowflake.cli._plugins.spcs.image_repository.commands.requests.get")
@mock.patch(
    "snowflake.cli._plugins.spcs.image_repository.manager.ImageRepositoryManager.execute_query"
)
@mock.patch(
    "snowflake.cli._plugins.spcs.image_repository.commands.ImageRepositoryManager._conn"
)
@mock.patch(
    "snowflake.cli._plugins.spcs.image_registry.manager.RegistryManager.login_to_registry"
)
def test_list_tags(
    mock_login,
    mock_conn,
    mock_execute,
    mock_get_tags,
    runner,
    mock_cursor,
):
    mock_conn.database = "DB"
    mock_conn.schema = "SCHEMA"
    mock_conn.role = "MY_ROLE"

    mock_execute.return_value = mock_cursor(rows=MOCK_ROWS_DICT, columns=MOCK_COLUMNS)
    mock_login.return_value = "TOKEN"

    mock_get_tags.return_value.status_code = 200
    mock_get_tags.return_value.text = (
        '{"name":"baserepo/super-cool-repo","tags":["1.2.0"]}'
    )

    result = runner.invoke(
        [
            "spcs",
            "image-repository",
            "list-tags",
            "IMAGES",
            "--image-name",
            "/DB/SCHEMA/IMAGES/super-cool-repo",
            "--format",
            "JSON",
        ]
    )

    assert result.exit_code == 0, result.output
    assert (
        "DeprecationWarning: The command 'list-tags' is deprecated."
        == result.output[0 : result.output.find("\n")]
    )
    assert json.loads(result.output[result.output.find("\n") :]) == [
        {"tag": "/DB/SCHEMA/IMAGES/super-cool-repo:1.2.0"}
    ], str(result.output)


@mock.patch(
    "snowflake.cli._plugins.spcs.image_repository.manager.ImageRepositoryManager.get_repository_url"
)
def test_get_repository_url_cli(mock_url, runner):
    repo_url = "repotest.registry.snowflakecomputing.com/db/schema/IMAGES"
    mock_url.return_value = repo_url
    result = runner.invoke(["spcs", "image-repository", "url", "IMAGES"])
    assert result.exit_code == 0, result.output
    assert result.output.strip() == repo_url


@mock.patch(
    "snowflake.cli._plugins.spcs.image_repository.manager.ImageRepositoryManager.show_specific_object"
)
def test_get_repository_url(mock_get_row):
    expected_row = MOCK_ROWS_DICT[0]
    mock_get_row.return_value = expected_row
    result = ImageRepositoryManager().get_repository_url(repo_name="IMAGES")
    mock_get_row.assert_called_once_with(
        "image repositories", "IMAGES", check_schema=True
    )
    assert isinstance(expected_row, Dict)
    assert "repository_url" in expected_row
    assert result == f"https://{expected_row['repository_url']}"


@mock.patch(
    "snowflake.cli._plugins.spcs.image_repository.manager.ImageRepositoryManager.show_specific_object"
)
def test_get_repository_url_no_scheme(mock_get_row):
    expected_row = MOCK_ROWS_DICT[0]
    mock_get_row.return_value = expected_row
    result = ImageRepositoryManager().get_repository_url(
        repo_name="IMAGES", with_scheme=False
    )
    mock_get_row.assert_called_once_with(
        "image repositories", "IMAGES", check_schema=True
    )
    assert isinstance(expected_row, Dict)
    assert "repository_url" in expected_row
    assert result == expected_row["repository_url"]


@mock.patch(
    "snowflake.cli._plugins.spcs.image_repository.manager.ImageRepositoryManager._conn"
)
@mock.patch(
    "snowflake.cli._plugins.spcs.image_repository.manager.ImageRepositoryManager.show_specific_object"
)
def test_get_repository_url_no_repo_found(mock_get_row, mock_conn):
    mock_get_row.return_value = None
    mock_conn.database = "DB"
    mock_conn.schema = "SCHEMA"
    with pytest.raises(ProgrammingError) as e:
        ImageRepositoryManager().get_repository_url(repo_name="IMAGES")
    assert (
        e.value.msg
        == "Image repository 'DB.SCHEMA.IMAGES' does not exist or not authorized."
    )
    mock_get_row.assert_called_once_with(
        "image repositories", "IMAGES", check_schema=True
    )


@mock.patch(
    "snowflake.cli._plugins.spcs.image_repository.manager.ImageRepositoryManager._conn"
)
def test_get_repository_url_no_database_provided(mock_conn):
    mock_conn.database = None
    with pytest.raises(DatabaseNotProvidedError):
        ImageRepositoryManager().get_repository_url("IMAGES")


@mock.patch(
    "snowflake.cli._plugins.spcs.image_repository.manager.ImageRepositoryManager._conn"
)
def test_get_repository_url_no_schema_provided(mock_conn):
    mock_conn.database = "DB"
    mock_conn.schema = None
    with pytest.raises(SchemaNotProvidedError):
        ImageRepositoryManager().get_repository_url("IMAGES")


@mock.patch("snowflake.connector.connect")
@pytest.mark.parametrize(
    "command, parameters",
    [
        ("list", []),
        ("list", ["--like", "PATTERN"]),
        ("drop", ["NAME"]),
    ],
)
def test_command_aliases(mock_connector, runner, mock_ctx, command, parameters):
    ctx = mock_ctx()
    mock_connector.return_value = ctx

    result = runner.invoke(["object", command, "image-repository", *parameters])
    assert result.exit_code == 0, result.output
    result = runner.invoke(
        ["spcs", "image-repository", command, *parameters], catch_exceptions=False
    )
    assert result.exit_code == 0, result.output

    queries = ctx.get_queries()
    assert queries[0] == queries[1]
