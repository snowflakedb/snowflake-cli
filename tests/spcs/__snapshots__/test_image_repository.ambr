# serializer version: 1
# name: test_create_cli
  '''
  +-----------------------------------------------------------+
  | key    | value                                            |
  |--------+--------------------------------------------------|
  | status | Image Repository TEST_REPO successfully created. |
  +-----------------------------------------------------------+
  
  '''
# ---
# name: test_create_cli_replace_and_if_not_exists_fails
  '''
  Usage: default spcs image-repository create [OPTIONS] NAME
  Try 'default spcs image-repository create --help' for help.
  +- Error ----------------------------------------------------------------------+
  | Parameters '--if-not-exists' and '--replace' are incompatible and cannot be  |
  | used simultaneously.                                                         |
  +------------------------------------------------------------------------------+
  
  '''
# ---
